File: clients/ios/Lextures/Core/Offline/OutboxStore.swift
import Foundation

/// Ordered write queue persisted locally until replay succeeds.
actor OutboxStore {
    private let fileURL: URL
    private var items: [OutboxItem] = []
    private var nextSequence = 0
    private var appliedKeys: Set<String> = []
    private let appliedKeysURL: URL

    init(ownerKey: String) {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Offline", isDirectory: true)
            .appendingPathComponent(ownerKey, isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        fileURL = base.appendingPathComponent("outbox.json")
        appliedKeysURL = base.appendingPathComponent("applied-idempotency-keys.json")
        load()
    }

    func enqueue(method: String, path: String, bodyJSON: String?, label: String) -> OutboxItem {
        let item = OutboxItem(
            id: UUID().uuidString.lowercased(),
            createdAt: Date(),
            sequence: nextSequence,
            method: method,
            path: path,
            bodyJSON: bodyJSON,
            label: label,
            status: .queued,
            lastError: nil
        )
        nextSequence += 1
        items.append(item)
        persist()
        return item
    }

    func pendingItems() -> [OutboxItem] {
        items.filter { $0.status == .queued || $0.status == .failed }
            .sorted { $0.sequence < $1.sequence }
    }

    func pendingCount() -> Int {
        items.filter { $0.status == .queued || $0.status == .failed || $0.status == .syncing }.count
    }

    func allItems() -> [OutboxItem] {
        items.sorted { $0.sequence < $1.sequence }
    }

    func update(_ item: OutboxItem) {
        guard let index = items.firstIndex(where: { $0.id == item.id }) else { return }
        items[index] = item
        persist()
    }

    func markApplied(idempotencyKey: String) {
        appliedKeys.insert(idempotencyKey)
        persistAppliedKeys()
    }

    func wasApplied(idempotencyKey: String) -> Bool {
        appliedKeys.contains(idempotencyKey)
    }

    func retry(id: String) {
        guard let index = items.firstIndex(where: { $0.id == id }) else { return }
        items[index].status = .queued
        items[index].lastError = nil
        persist()
    }

    func clearAll() {
        items.removeAll()
        nextSequence = 0
        appliedKeys.removeAll()
        persist()
        persistAppliedKeys()
    }

    // MARK: - Private

    private struct PersistedOutbox: Codable {
        var items: [OutboxItem]
        var nextSequence: Int
    }

    private func load() {
        if
            let data = try? Data(contentsOf: fileURL),
            let decoded = try? JSONDecoder().decode(PersistedOutbox.self, from: data) {
            items = decoded.items
            nextSequence = decoded.nextSequence
        }
        if
            let data = try? Data(contentsOf: appliedKeysURL),
            let keys = try? JSONDecoder().decode(Set<String>.self, from: data) {
            appliedKeys = keys
        }
    }

    private func persist() {
        let payload = PersistedOutbox(items: items, nextSequence: nextSequence)
        guard let data = try? JSONEncoder().encode(payload) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }

    private func persistAppliedKeys() {
        guard let data = try? JSONEncoder().encode(appliedKeys) else { return }
        try? data.write(to: appliedKeysURL, options: .atomic)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/offline/OutboxStore.kt
package com.lextures.android.core.offline

import android.content.Context
import java.io.File
import java.util.UUID
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
private data class PersistedOutbox(
    val items: List<OutboxItem> = emptyList(),
    val nextSequence: Int = 0,
)

class OutboxStore(context: Context, ownerKey: String) {
    private val rootDir = File(context.applicationContext.filesDir, "offline/$ownerKey").apply { mkdirs() }
    private val outboxFile = File(rootDir, "outbox.json")
    private val appliedKeysFile = File(rootDir, "applied-idempotency-keys.json")
    private val json = Json { ignoreUnknownKeys = true }
    private var state = loadOutbox()
    private var appliedKeys: MutableSet<String> = loadAppliedKeys()

    fun enqueue(method: String, path: String, bodyJson: String?, label: String): OutboxItem {
        val item = OutboxItem(
            id = UUID.randomUUID().toString(),
            sequence = state.nextSequence,
            method = method,
            path = path,
            bodyJson = bodyJson,
            label = label,
        )
        state = state.copy(
            items = state.items + item,
            nextSequence = state.nextSequence + 1,
        )
        persistOutbox()
        return item
    }

    fun pendingItems(): List<OutboxItem> =
        state.items.filter {
            it.outboxStatus() == OutboxStatus.Queued || it.outboxStatus() == OutboxStatus.Failed
        }.sortedBy { it.sequence }

    fun pendingCount(): Int =
        state.items.count {
            val status = it.outboxStatus()
            status == OutboxStatus.Queued || status == OutboxStatus.Failed || status == OutboxStatus.Syncing
        }

    fun allItems(): List<OutboxItem> = state.items.sortedBy { it.sequence }

    fun update(item: OutboxItem) {
        state = state.copy(items = state.items.map { if (it.id == item.id) item else it })
        persistOutbox()
    }

    fun markApplied(idempotencyKey: String) {
        appliedKeys.add(idempotencyKey)
        appliedKeysFile.writeText(json.encodeToString(appliedKeys))
    }

    fun wasApplied(idempotencyKey: String): Boolean = appliedKeys.contains(idempotencyKey)

    fun retry(id: String) {
        state = state.copy(
            items = state.items.map { item ->
                if (item.id == id) item.copy(status = OutboxStatus.Queued.name, lastError = null) else item
            },
        )
        persistOutbox()
    }

    fun clearAll() {
        state = PersistedOutbox()
        appliedKeys.clear()
        persistOutbox()
        appliedKeysFile.delete()
    }

    private fun loadOutbox(): PersistedOutbox {
        if (!outboxFile.exists()) return PersistedOutbox()
        return runCatching { json.decodeFromString<PersistedOutbox>(outboxFile.readText()) }
            .getOrDefault(PersistedOutbox())
    }

    private fun persistOutbox() {
        outboxFile.writeText(json.encodeToString(state))
    }

    private fun loadAppliedKeys(): MutableSet<String> {
        if (!appliedKeysFile.exists()) return mutableSetOf()
        return runCatching { json.decodeFromString<Set<String>>(appliedKeysFile.readText()).toMutableSet() }
            .getOrDefault(mutableSetOf())
    }
}
